src/main.rs

```diff
@@ -130,144 +136,129 @@ fn main() {
-                // So, we're handling a Broken Pipe error and exiting with 0 anyway
-                let stdout = std::io::stdout();
-                {
-                    let mut handle = stdout.lock();
-                    handle.write_all(line.as_bytes()).unwrap_or_else(|e| {
-                        match e.kind() {
-                            std::io::ErrorKind::BrokenPipe => std::process::exit(0),
-                            _ => std::process::exit(1),
-                        };
-                    });
+                if solve_cond && (filter_cond || subargs.filter.is_none()) {
+                    let line = if subargs.paths {
+                        format!("{}\n", fname)
+                    } else if subargs.names {
+                        format!("{}\n", e.name)
+                    } else {
+                        format!("{:<17}\t{:<46}\t{:<7}\n", e.name, fname, status)
+                    };
+                    // Somehow using println! leads to the binary panicking
+                    // when its output is piped.
+                    // So, we're handling a Broken Pipe error and exiting with 0 anyway
+                    let stdout = std::io::stdout();
+                    {
+                        let mut handle = stdout.lock();
+                        handle.write_all(line.as_bytes()).unwrap_or_else(|e| {
+                            match e.kind() {
+                                std::io::ErrorKind::BrokenPipe => std::process::exit(0),
+                                _ => std::process::exit(1),
+                            };
+                        });
+                    }
                 }
-            }
-        });
-        let percentage_progress = exercises_done as f32 / exercises.len() as f32 * 100.0;
-        println!(
-            "Progress: You completed {} / {} exercises ({:.2} %).",
-            exercises_done,
-            exercises.len(),
-            percentage_progress
-        );
-        std::process::exit(0);
-    }
-
-    // Handle the run command
-    if let Some(ref matches) = matches.subcommand_matches("run") {
-        let name = matches.value_of("name").unwrap();
-
-        let matching_exercise = |e: &&Exercise| name == e.name;
-
-        let exercise = exercises.iter().find(matching_exercise).unwrap_or_else(|| {
-            println!("No exercise found for your given name!");
-            std::process::exit(1)
-        });
+            });
+            let percentage_progress = exercises_done as f32 / exercises.len() as f32 * 100.0;
+            println!(
+                "Progress: You completed {} / {} exercises ({:.2} %).",
+                exercises_done,
+                exercises.len(),
+                percentage_progress
+            );
+            std::process::exit(0);
+        }
 
-        run(&exercise, verbose).unwrap_or_else(|_| std::process::exit(1));
-    }
+        Subcommands::Run(subargs) => {
+            let exercise = find_exercise(subargs.name, exercises);
```

Comment:
Borrow parameters if `find_exercise` signature changes.
```suggestion
            let exercise = find_exercise(&subargs.name, &exercises);
```